src/enrichment_tables/geoip.rs

```diff
@@ -0,0 +1,429 @@
+// products. If we encounter one of these two types, we look for ASN/ISP information;
+// otherwise we expect to be working with a City database.
+const ASN_DATABASE_TYPE: &str = "GeoLite2-ASN";
+const ISP_DATABASE_TYPE: &str = "GeoIP2-ISP";
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
+pub struct GeoipConfig {
+    pub path: String,
+    #[serde(default = "default_locale")]
+    pub locale: String,
+}
+
+// valid locales are: “de”, "en", “es”, “fr”, “ja”, “pt-BR”, “ru”, and “zh-CN”
+//
+// https://dev.maxmind.com/geoip/docs/databases/city-and-country?lang=en
+//
+// TODO try to determine the system locale and use that as default if it matches a valid locale?
+fn default_locale() -> String {
+    "en".to_string()
+}
+
+impl GenerateConfig for GeoipConfig {
+    fn generate_config() -> toml::Value {
+        toml::Value::try_from(Self {
+            path: "/path/to/GeoLite2-City.mmdb".to_string(),
+            locale: default_locale(),
+        })
+        .unwrap()
+    }
+}
+
+inventory::submit! {
+    EnrichmentTableDescription::new::<GeoipConfig>("geoip")
+}
+
+#[async_trait::async_trait]
+#[typetag::serde(name = "geoip")]
+impl EnrichmentTableConfig for GeoipConfig {
+    async fn build(
+        &self,
+        _: &crate::config::GlobalOptions,
+    ) -> crate::Result<Box<dyn Table + Send + Sync>> {
+        Ok(Box::new(Geoip::new(self.clone())?))
+    }
+}
+
+#[derive(Clone)]
+pub struct Geoip {
+    config: GeoipConfig,
+    dbreader: Arc<maxminddb::Reader<Vec<u8>>>,
+    last_modified: SystemTime,
+}
+
+impl Geoip {
+    pub fn new(config: GeoipConfig) -> crate::Result<Self> {
+        let table = Geoip {
+            last_modified: fs::metadata(&config.path)?.modified()?,
+            dbreader: Arc::new(Reader::open_readfile(config.path.clone())?),
+            config,
+        };
+
+        // Check if we can read database with dummy Ip.
+        let ip = IpAddr::V4(std::net::Ipv4Addr::new(0, 0, 0, 0));
+        let result = if table.has_isp_db() {
+            table.dbreader.lookup::<Isp>(ip).map(|_| ())
+        } else {
+            table.dbreader.lookup::<City>(ip).map(|_| ())
+        };
+
+        match result {
+            Ok(_) | Err(MaxMindDBError::AddressNotFoundError(_)) => (),
+            Err(error) => return Err(error.into()),
+        }
+
+        Ok(table)
+    }
+
+    fn has_isp_db(&self) -> bool {
+        self.dbreader.metadata.database_type == ASN_DATABASE_TYPE
+            || self.dbreader.metadata.database_type == ISP_DATABASE_TYPE
+    }
+
+    fn lookup(&self, ip: IpAddr, select: Option<&[String]>) -> Option<BTreeMap<String, Value>> {
+        let mut map = BTreeMap::new();
+        let mut add_field = |key: &str, value: Option<Value>| {
+            if select
+                .map(|fields| fields.iter().any(|field| field == key))
+                .unwrap_or(true)
+            {
+                map.insert(key.to_string(), value.unwrap_or(Value::Null));
+            }
+        };
```

Comment:
Should we convert the `select` here to a `HashSet` to make the lookups cheaper?

Reply:
For somewhat larger `select` lookups would be cheaper, but considering small size of `select` and the cost of creating `HashSet` just iterating over it should be slightly faster or comparable to using `HashSet`.